apps/passport-client/pages/index.tsx

```diff
@@ -28,30 +34,38 @@ import {
   saveIdentity
 } from "../src/localstorage";
 import { registerServiceWorker } from "../src/registerServiceWorker";
-import { AppState } from "../src/state";
+import { AppState, StateEmitter } from "../src/state";
 import { pollUser } from "../src/user";
 
 class App extends React.Component<object, AppState> {
   state = undefined as AppState | undefined;
+  stateEmitter: StateEmitter = new Emitter();
   update = (diff: Pick<AppState, keyof AppState>) => {
     console.log("App.update", diff);
-    this.setState(diff);
+    this.setState(diff, () => {
+      this.stateEmitter.emit(this.state);
+    });
   };
   dispatch = (action: Action) => dispatch(action, this.state, this.update);
   componentDidMount() {
     loadInitialState().then((s) => this.setState(s, this.startBackgroundJobs));
   }
+  stateContextState: StateContextState = {
+    getState: () => this.state,
+    stateEmitter: this.stateEmitter,
+    dispatch: this.dispatch
+  };
 
   render() {
-    const { state, dispatch: disp } = this;
+    const { state } = this;
 
     if (!state) {
       return null;
     }
 
     const hasStack = state.error?.stack != null;
     return (
-      <DispatchContext.Provider value={[state, disp]}>
+      <StateContext.Provider value={this.stateContextState}>
```

Comment:
now, we pass a stable object into this global provider, and its consumers now explicitly subscribe to re-render depending on whether a particular sub-property of this object changes, rather than on any change to any property of this global state object